metrics/metrics.go

```diff
@@ -0,0 +1,167 @@
+
+	"github.com/letsencrypt/boulder/rpc"
+)
+
+// HTTPMonitor stores some server state
+type HTTPMonitor struct {
+	stats           statsd.Statter
+	statsPrefix     string
+	handler         http.Handler
+	openConnections *int64
+}
+
+// NewHTTPMonitor returns a new initialized HTTPMonitor
+func NewHTTPMonitor(stats statsd.Statter, handler http.Handler, prefix string) HTTPMonitor {
+	conns := int64(0)
+	return HTTPMonitor{stats: stats, handler: handler, statsPrefix: prefix, openConnections: &conns}
+}
+
+// Handle wraps handlers and records various metrics about requests to these handlers
+// and sends them to StatsD
+func (h *HTTPMonitor) Handle() http.Handler {
+	return http.HandlerFunc(h.watchAndServe)
+}
+
+func (h *HTTPMonitor) watchAndServe(w http.ResponseWriter, r *http.Request) {
+	h.stats.Inc(fmt.Sprintf("%s.HTTP.Rate", h.statsPrefix), 1, 1.0)
+	open := atomic.AddInt64(h.openConnections, 1)
+	h.stats.Gauge(fmt.Sprintf("%s.HTTP.OpenConnections", h.statsPrefix), open, 1.0)
+
+	cOpened := time.Now()
+	h.handler.ServeHTTP(w, r)
+	cClosed := time.Since(cOpened)
+
+	open = atomic.AddInt64(h.openConnections, -1)
+	h.stats.Gauge(fmt.Sprintf("%s.HTTP.OpenConnections", h.statsPrefix), open, 1.0)
+
+	// Check if request failed
+	state := "Success"
+	if w.Header().Get("Content-Type") == "application/problem+json" {
+		state = "Error"
+	}
+
+	// If r.URL has more than two segments throw the rest away to simplify metrics
+	segments := strings.Split(r.URL.Path, "/")
+	if len(segments) > 3 {
+		segments = segments[:3]
+	}
+	endpoint := strings.Join(segments, "/")
+
+	h.stats.TimingDuration(fmt.Sprintf("%s.HTTP.ResponseTime.%s.%s", h.statsPrefix, endpoint, state), cClosed, 1.0)
+}
+
+// RPCMonitor stores rpc delivery state
+type RPCMonitor struct {
+	deliveryTimings map[string]time.Time
+	dtMu            *sync.Mutex
+
+	stats statsd.Statter
+}
+
+// NewRPCMonitor returns a new initialized RPCMonitor and starts a goroutine
+// to cleanup timeouts from the delivery map
+func NewRPCMonitor(stats statsd.Statter) RPCMonitor {
+	r := RPCMonitor{
+		stats:           stats,
+		deliveryTimings: make(map[string]time.Time),
+		dtMu:            &sync.Mutex{},
+	}
+	go func() {
+		c := time.Tick(time.Second * 5)
+		for _ = range c {
+			if t := r.cleanup(); t > 0 {
+				stats.Inc("RPC.Timeouts", t, 1.0)
+			}
+		}
+	}()
+	return r
+}
+
+func (r *RPCMonitor) size() int {
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	return len(r.deliveryTimings)
+}
+
+func (r *RPCMonitor) get(id string) time.Time {
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	return r.deliveryTimings[id]
+}
+
+func (r *RPCMonitor) add(id string) {
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	r.deliveryTimings[id] = time.Now()
+}
+
+func (r *RPCMonitor) delete(id string) {
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	delete(r.deliveryTimings, id)
+}
+
+func (r *RPCMonitor) cleanup() (removed int64) {
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	for k, v := range r.deliveryTimings {
```

Comment:
This seems like a long time to have the lock esp if there are lots of requests.

At the very least, that time.Now().Add() calculation only needs to be done once and can be done outside the lock.

It feels like there is a smarter way of doing this but I don't have an immediate solution. Have you looked at what similar libraries do, like go-metrics or github.com/codahale/metrics? They might be very simple, too, but maybe have a good idea in them.